CSID-460: AKS tutorials: aks-base and replicator-aks-cc

gianlucanatali opened the pull request:
### Description 

As a Developer, DevOps, Sales Engineer, Microsoft SE, I want to have a guided tutorial/example configure Confluent Platform running on Azure Kubernetes Service (AKS) that also include how to replicate data to Confluent Cloud running in Azure , so that I can easily demonstrate the value of running Confluent Platform in Azure both self manage and on prem, and the ease of use of leveraging Operator with AKS

https://confluentinc.atlassian.net/browse/CSID-460

_What behavior does this PR change, and why?_
This will add tutorials for AKS (Azure Kubernetes Service) , reusing the gke tutorials common code

### Author Validation

_Describe the validation already done, or needs to be done, by the PR submitter._

This is a new sample, so it should not change existing behaviors

<!-- Uncomment any of the following that are required -->
- [ ] Documentation 
<!-- - [ ] ccloud -->
<!-- - [ ] ccloud/beginner-cloud -->
<!-- - [ ] ccloud/ccloud-stack -->
<!-- - [ ] clickstream -->
<!-- - [ ] clients/avro -->
<!-- - [ ] clients/cloud -->
<!-- - [ ] cloud-etl -->
<!-- - [ ] connect-streams-pipeline -->
<!-- - [ ] cp-quickstart -->
<!-- - [ ] kubernetes/gke-base -->
<!-- - [ ] kubernetes/replicator-gke-cc -->
<!-- - [ ] microservices-orders -->
<!-- - [ ] multi-datacenter -->
<!-- - [ ] multiregion -->
<!-- - [ ] music -->
<!-- - [ ] replicator-schema-translation -->
<!-- - [ ] replicator-security -->
<!-- - [ ] security/rbac -->
<!-- - [ ] security/secret-protection -->


### Reviewer Tasks

@rspurgeon 

_Describe the tasks/validation that the PR submitter is requesting to be done by the reviewer._

<!-- Uncomment any of the following that are required -->
- [ ] Documentation 

<!-- - [ ] ccloud -->
<!-- - [ ] ccloud/beginner-cloud -->
<!-- - [ ] ccloud/ccloud-stack -->
<!-- - [ ] clickstream -->
<!-- - [ ] clients/avro -->
<!-- - [ ] clients/cloud -->
<!-- - [ ] cloud-etl -->
<!-- - [ ] connect-streams-pipeline -->
<!-- - [ ] cp-quickstart -->
<!-- - [ ] kubernetes/gke-base -->
<!-- - [ ] kubernetes/replicator-gke-cc -->
<!-- - [ ] microservices-orders -->
<!-- - [ ] multi-datacenter -->
<!-- - [ ] multiregion -->
<!-- - [ ] music -->
<!-- - [ ] replicator-schema-translation -->
<!-- - [ ] replicator-security -->
<!-- - [ ] security/rbac -->
<!-- - [ ] security/secret-protection -->

Docs Repo has the staged docs on PR 192

gianlucanatali committed:
CSID-460: Cloning gke-base as is

gianlucanatali committed:
CSID-460: First draft of AKS k8s example

gianlucanatali committed:
CSID-460: Fixing mlformed table in docs

gianlucanatali committed:
CSID-460: Fixing mlformed table in docs 857e429

gianlucanatali committed:
CSID-460: Fixing doc double reference

gianlucanatali committed:
CSID-460: Removing un-needed line

gianlucanatali committed:
CSID-460: Changing labels to avoid duplicates

gianlucanatali committed:
CSID-460: Changing labels to avoid duplicates

gianlucanatali committed:
CSID-460: trying changing label name

gianlucanatali committed:
CSID-460: Adding instruction to set up azure cli before running example

gianlucanatali committed:
CSID-460: Using newer k8s version avail in Azure

gianlucanatali committed:
CSID-460: Creating a clone of replicator-gke-cc

gianlucanatali committed:
CSID-460: Fist draft replicator-aks-cc

gianlucanatali committed:
CSID-460: Fixing malformed table

gianlucanatali committed:
CSID-460: Removing title from included rst

gianlucanatali committed:
CSID-460: Adding Title after label

gianlucanatali committed:
CSID-460: Creating ccloud config details for Azure

gianlucanatali committed:
CSID-460: Removing refuse gke from aks ccloud docs

gianlucanatali committed:
CSID-460: Adding output type to az aks list command in documentation

gianlucanatali committed:
CSID-460: fix variable name ADDITIONAL_HELM_FLAGS

gianlucanatali committed:
CSID-460: fixing pdb.yaml file

The pdb.yaml file for kafka deployment template is using hooks that are not needed for k8s version higher than 1.15+ . As this demo will use a newer version I need to comment out 2 lines of the pdb.yaml file. I am doing this by replacing the default file with the modified one as part of the make demo

gianlucanatali committed:
CSID-460: Using --no-hooks helm param

gianlucanatali committed:
CSID-460: Updating architecture image

gianlucanatali committed:
CSID-460: Changing default vm_type for aks

gianlucanatali committed:
CSID-460: fix malformed table

gianlucanatali committed:
CSID-460: fixing rst issue

rspurgeon commented:
Looking good so far @gianlucanatali ... FYI when I run `make demo` I do get redundant validations of installed tools:

```
 make demo
✔ found curl
✔ found tar
✔ found kubectl
✔ found helm
✔ found jq
✔ found az
✔ aks-base dependencies verified
✔ Confluent Platform Operator helm bundle ready at: /Users/rspurgeon/dev/confluentinc/examples/kubernetes/common/cp/operator/1.6.0/
✔ common initialization complete

✔ aks-base demo validation complete
✔ found curl
✔ found tar
✔ found kubectl
✔ found helm
✔ found jq
✔ Confluent Platform Operator helm bundle ready at: /Users/rspurgeon/dev/confluentinc/examples/kubernetes/common/cp/operator/1.6.0/
✔ common initialization complete
...
```

gianlucanatali commented on kubernetes/aks-base/Makefile-impl:
```diff
@@ -0,0 +1,67 @@
+	kubectl --context $(KUBECTL_CONTEXT) -n $(KUBECTL_NAMESPACE) apply -f $(AKS_BASE_MKFILE_DIR)cfg/client-console-pod.yaml
+	@$(call echo_stdout_footer_pass,client-console deployed)
+
+aks-base-validate: aks-check-dependencies init
```
Hi @rspurgeon , I think this line is causing the double check of the dependencies.
I have checked in the gke version the **init** is not there , so probably needs to be removed.
The init will be called by base-demo-validate <- run-base-demo  <- aks-base-demo (line 57 in this file)

gianlucanatali commented on kubernetes/replicator-aks-cc/Makefile-impl:
```diff
@@ -0,0 +1,44 @@
+	@$(call echo_stdout_footer_pass,client-console deployed)
+##################################################################
+
+replicator-aks-cc-demo:
```
I think we should be calling aks-base-validate in this function, otherwise those checks are not made AFAICT
We'd also need to add gke-base-validate to the replicator-gke-cc-demo
What do you think @rspurgeon ?

gianlucanatali commented on kubernetes/aks-base/Makefile-impl:
```diff
@@ -0,0 +1,67 @@
+ 
+HELM_COMMON_FLAGS := --wait --timeout=5m -f $(AKS_BASE_MKFILE_DIR)cfg/values.yaml --set global.initContainer.image.tag=$(OPERATOR_CP_IMAGE_TAG) --set global.provider.region=$(AKS_BASE_REGION) --set global.provider.kubernetes.deployment.zones={$(AKS_BASE_REGION)-$(AKS_BASE_ZONE)} $(ADDITIONAL_HELM_FLAGS)
+
+aks-check-dependencies: check-dependencies
```
this is another offender , **check-dependencies** must go as it is already in the **init** function
@rspurgeon

gianlucanatali committed:
CSID-460: Fixing double dependencies check for aks-base

gianlucanatali committed:
CSID-460: Adding cloud provider clients validation on replicator demos

I am also fixing the GKE demo, as it had the same shortcoming.
This would cause the demo to fail if the az client is not installed or if the resource_group variable is not defined,  but without a proper error message as it is not checked.
Similar issue for GKE
@rspurgeon

gianlucanatali commented:
Thanks @rspurgeon , I think I found the culprit!
I have fixed both issues, lmk what you think 👍

rspurgeon commented:
> Thanks @rspurgeon , I think I found the culprit!
> I have fixed both issues, lmk what you think 👍

Yes, that seems to have removed the redundancy.  Thanks

rspurgeon commented:
@gianlucanatali I was able to verify both demos and scan the documents.  I say we merge this in.  Go ahead and move it to "Ready for review"

gianlucanatali commented:
Hi @rspurgeon , Now the PR is ready for review. Thanks a lot!

rspurgeon reviewed (approved):
LGTM, nice addition @gianlucanatali
